[2기-P] 김민성 SprintBoot URL Shortener 구현 미션 제출합니다.

epicblues opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

기본 요구사항만 구현했습니다! 패키지 구성에 고민이 많았습니다

- [x]  URL 입력폼 제공 및 결과 출력

```bash
GET "/" : URL 입력폼 페이지
```
- [x]  URL Shortening Key는 8 Character 이내로 생성

- [x] 단축된 URL 요청시 원래 URL로 리다이렉트

- [x] 단축된 URL에 대한 요청 수 정보저장 (optional)
```bash
GET "/meta" : 단축 URL 요청 수 출력 페이지
```
- [ ] Shortening Key를 생성하는 알고리즘 2개 이상 제공하며 애플리케이션 실행중 동적으로 변경 가능 (미구현)


## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

1. 단순한 html 페이지를 반환하는 API도 명세 작성 대상에 포함되는지 궁금합니다! 축약된 url이 존재하지 않을 경우 실패 페이지 html 뷰를 응답하는 API를 작성할 때 이런 의구심이 들었습니다.

2. 패키지 구조를 비즈니스 로직이 담긴 domain과 http API를 담당하는 presentation으로 크게 나누었습니다.  과제를 진행하면서 DB와 통신하는 클래스(ShortenedUrlRepository와 구현체들)도  따로 패키지(dao?)로 묶어서 관리했어야 했을까요?

3. 멘토님은 인터페이스에 대한 구현체를 인터페이스와 같은 패키지에 넣어서 관리하시는지 궁금합니다. 멘토님이 추천하신 [toss 발표 영상](https://www.youtube.com/watch?v=RVO02Z1dLF8&list=PL1DJtS1Hv1PiGXmgruP1_gM2TSvQiOsFL&t=539s)에서 implementation 계층을 따로 구성한 것에 영감(?)을 받아서 도메인의 주요 인터페이스에 대한 구현 클래스들을 하위 패키지인 implementation에 배치했는데, 차라리 domain과 독립적인 implementation 패키지를 만드는 것이 더 나았을까요? 

4. HttpController 관련 예외 처리만 담당하는 ControllerAdvice인 클래스 이름을 HttpExceptionHandler라 정하고, 사용하는 예외들과 함께 exception 패키지에 넣어놨는데, 클래스 이름과 패키지 위치가 멘토님이 느끼기에 부자연스러운 부분이 있는지 궁금합니다!  HttpController와 같은 패키지에 배치하는 게 더 나았을까요?

5. [comment](https://github.com/prgrms-be-devcourse/springboot-url-shortener/pull/8#discussion_r898735158) 확인 부탁드립니다!

6. 멘토님은 테스트 코드를 nested class를 활용해서 계층화하여 작성하시는지 궁금합니다! 만약 사용하신다면 어떤 목적으로 사용하시는지도 알고 싶습니다✍️

epicblues committed:
chore : 프로젝트 기본 설정

의존성 설정

epicblues committed:
feat : 도메인 추가

* ShortenedUrl, ShortenedUrlTest
프로젝트 핵심 도메인
DB에서 자동으로 생성되는 id 보관
사용자가 줄이기를 원하는 원본 URL 보관
id 값을 해싱한 키 값 보관

epicblues committed:
feat : URL 축약 문자열 생성 서비스

* ShortenedUrlService, ShortenedUrlServiceTest
url 저장 유무 확인
이미 url이 존재할 경우 그 url의 key값 반환
없을 경우 새로 url 만들고 해당하는 key값 반환

epicblues committed:
feat : Jpa 의존 ShortenedUrl 생성 모듈

* JpaShortenedUrlFactory, JpaShortenedUrlFactoryTest
jpa repository api를 직접 사용해서 ShortendUrl entity를 생성
축약된 key를 만드는 기능은 UrlShorteningStrategy에 위임

* UrlShorteningStrategy
문자열 축약 기능을 수행하는 인터페이스

* JpaShortenedUrlRepository
ShortenedUrlRepository의 JPA 구현체 버전

epicblues committed:
test : JpaShortenedUrlRepository

상위 인터페이스 ShortenedUrlRepository의 api 정상 구현 및 작동 확인

epicblues committed:
refactor : 테스트 코드

epicblues committed:
fix : 문자열 길이 요구사항

epicblues committed:
feat : Base62 기본 문자열 인코딩

* Base62EncodingStrategy, Base62EncodingStrategyTest
1 ~ 2147483646 까지 인코딩(Integer 한계)

* ShortenedUrl
id 값 정수 범위 Integer로 변경

epicblues committed:
chore : 패키지 구조 변경

epicblues committed:
feat : url 축약 Http API

* HttpController, HttpControllerTest
POST "/url"

epicblues committed:
feat : 축약 url 매핑 http api

* HttpController, HttpControllerTest
GET /{key}
축약된 url과 매핑되는 원본 url이 존재할 경우 원본 url로 redirect
그렇지 않을 경우 404 Not Found 응답과 html 화면 출력

ShortenedUrlNotFoundException, HttpControllerAdvice
* 축약된 url 관련 정보를 db에서 찾지 못했을 때 던지는 예외
* 해당 예외 처리 로직(404 Not Found + failure.html view)

ShortenedUrlService, ShortenedUrlRepository
* key 값을 가진 ShortenedUrl 엔티티 검색 및 원본 url 반환

epicblues committed:
feat : 메인 HTML 페이지

ajax 요청을 통해 축약된 url을 서버로 부터 받는 페이지

* pom.xml
thymeleaf 추가

* JpaShortenedUrlRepositoryTest
축약된 key를 통한 조회
JPA 구현체 기능 동작 여부 테스트

* ShortenedUrlPayload
HttpConverter가 제대로 작동하기 위한 기본 생성자
필드 값 final 제거(기본 생성자를 사용할 수 없게 된다.)

epicblues committed:
feat : HTTP API 잘못된 url 입력 예외

* POST /url
InvalidUrlRequestException
URL 이 http URL 형식에 맞지 않을 경우 클라이언트에게 400 응답

epicblues committed:
refactor : 예외 처리 패키지 구조 변경

HttpExceptionHandler -> exception 패키지로 이동

epicblues committed:
refactor : 테스트 코드 패키지 구조

epicblues committed:
feat : 예상치 못한 예외 처리 응답

* HttpExceptionHandler, MockHttpControllerTest
컨트롤러 계층에서 따로 처리하지 못하는 예외가 발생했을 때 500 Internal Server Error 상태 코드를 반환하는 api 설정

epicblues committed:
chore : DB 환경 분리

테스트용 환경(메모리 DB)과 실제 애플리케이션 실행 환경(파일 기반 DB) 분리

blessing333 commented on src/main/java/com/prgrms/shortener/presentation/HttpController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+    String key = shortenedUrlService.shorten(payload.getUrl());
+    String requestUrl = httpServletRequest.getRequestURL().toString();
+    String urlWithoutPath = requestUrl.substring(0, requestUrl.length() - 3);
```
이 부분은 localhost:8080 호스트 주소를 추출하는 로직일까요?
만약 그렇다면 모든 요청마다 호스트 주소를 계산하지 말고 저장해두고 사용하는건 어떨까용
문자열 상수나 apllication.yml에 app.host로 저장해놓고 사용하는 방법이 있을것같아요!

epicblues replied:
이건 제가 예전 프로젝트에서 AWS EC2의 인스턴스를 다시 시작할 때마다 호스트 주소가 새로 할당되기 때문에 매번 변경된 호스트 주소를 수동으로  수정하는 것이 번거로웠던 악몽(?) 때문에 작성한 코드입니다. 아마 호스트 주소가 거의 고정된 상황이라면 말씀하신 방법을 사용할 것 같습니다..! 
아니면 제 해결책에서 잠재적으로 발생할 수 있는 문제가 있을까요?

blessing333 commented on src/main/java/com/prgrms/shortener/presentation/HttpController.java:
```diff
@@ -0,0 +1,51 @@
+
+@Controller
+public class HttpController {
+
```
HttpController 라는 클래스 이름은 살짝 모호한것 같아요!  json응답을 반환하는 부분과 html 뷰를 반환하는 부분을 나누는건 어떨까요?? (사실 크기가 너무 작아 애매한것 같기도 합니다)

epicblues replied:
초반에 과제를 수행할 때는 컨트롤러의 메서드가 많지 않을 줄 알고(!) 컨트롤러 하나에서 다 처리하는 것이 오히려 바람직하다고 생각했는데,  말씀하신대로 클래스 이름도 모호하고, 코드를 작성하면서 `@ResponseBody` 를 까먹는 실수도 자주 발생했기 때문에  바꿔야할 것 같습니다. 🙄

blessing333 commented on src/main/java/com/prgrms/shortener/presentation/HttpController.java:
```diff
@@ -0,0 +1,51 @@
+  public String handleShortenedUrl(@PathVariable String key) {
+
+    Optional<String> originalUrl = shortenedUrlService.findOriginalUrlByKey(key);
+
```
개인적인 생각입니다.
패키지 구조를 보면 내부 영역(domain)과 외부 영역(presentation)을 나누신것같습니다.
이런 경우 ShortenedUrlNotFoundException은 내부 영역(도메인 서비스) 계층에 위치하는게 자연스러울것 같다는 생각이 들었습니다
shortenedUrlService.findOriginalUrlByKey(key) 메서드 내부에서 ShortenedUrlNotFoundException 예외를 다루는건 어떨까요??

epicblues replied:
안 그래도 지금 방식으로는 stack trace가 컨트롤러에서 시작해서 느낌이 안좋았는데, 차라리 도메인 영역으로 옮겨서 처리하는게 낫겠네요!

blessing333 commented on src/main/java/com/prgrms/shortener/domain/implementation/Base62EncodingStrategy.java:
```diff
@@ -0,0 +1,24 @@
+import com.prgrms.shortener.domain.encoding.EncodingStrategy;
+import org.springframework.stereotype.Component;
+
+@Component
```
이 친구는 encoding 패키지에 넣는건 어떤가용??

epicblues replied:
toss 영상의 implementation layer에 혹해서 과감하게 시도해봤는데 망한 것 같습니다.  ☠️ 민재님과 멘토님의 피드백을 반영해서 수정할 것 같습니다.

blessing333 commented on src/main/java/com/prgrms/shortener/domain/implementation/Base62EncodingStrategy.java:
```diff
@@ -0,0 +1,24 @@
+public class Base62EncodingStrategy implements EncodingStrategy {
+
+  private final String[] table = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".split("");
+
```
이렇게 하니까 가독성이 훨씬 좋아지는군요

blessing333 commented on src/main/java/com/prgrms/shortener/presentation/validation/HttpUrlValidator.java:
```diff
@@ -0,0 +1,17 @@
+  public void initialize(HttpURL constraintAnnotation) {
+    ConstraintValidator.super.initialize(constraintAnnotation);
+  }
+
```
혹시 자바 validation에서 제공하는 @Url을 사용안하시고 별도의 커스텀 어노테이션을 만드신 이유가 있으신가요??

epicblues replied:
학습 차원에서 시도해봤습니다.  실제 프로젝트에는  `@URL` 을 공부해서 사용할 것 같습니다...!

blessing333 reviewed (commented):
테스트 커버리지 98%... 경이롭네요 👍 
3번 질문과 4번 질문은 저도 궁금하네요
(여기서부터는 지극히 개인적인 의견입니다!)
상황에 따라 다르겠지만 저는 일단 인터페이스와 구현체는 최대한 가까이에 두는게 자연스럽다고 생각합니다. implementation 패키지에 있는 Base62Encoding 구현체가 encoding 패키지에 있는게 더 이해하기 쉬울것 같아요.
다만 인터페이스의 구현체가 외부 라이브러리일 경우에는 도메인 영역 외부에 두는게 좋을것 같아요.
예를 들어 ShortenedUrlRepository는 도메인 패키지 내부에 두고, 그 구현체인 JpaShortenedUrlRepository는 도메인 외부 별도의 패키지(infra)에 작성할 것 같습니다.

epicblues committed:
feat : 축약 url 방문 시 조회수 갱신

* ShortenedUrl, ShortenedUrlTest
조회수 필드(count) 추가

* ShortenedUrlRepository, JpaShortenedUrlRepository, JpaShortenedUrlRepositoryTest
엔티티의 최신 조회수를 1 증가시키는 공개 API 추가.

* ShortenedUrlService, ShortenedUrlServiceTest
key 값을 통한 조회가 성공할 시에 조회수를 1 증가시킨 뒤 원본 url을 반환하도록 변경

epicblues committed:
feat : IllegalArgumentException 예외 처리 응답

* HttpExceptionHandler, HttpControllerTest
애플리케이션에서 발생하는 IllegalArgumentException 예외 처리 및 응답(400)

* ShortenedUrlService
원본 URL 요청에 대한 응답 Optional 제거
ShortenedUrlNotFoundException 예외 발생 방식으로 변경

* ShortenedUrlNotFoundException
presentation -> domain 패키지로 이동

epicblues committed:
refactor : Controller 분리

* HttpRestController, HttpController
JSON 관련 요청 처리 -> HttpRestController(advice : HttpRestExceptionHandler)
웹 페이지 관련 요청 처리 -> HttpController(advice : HttpExceptionHandler)

epicblues committed:
feat : 축약 URL 접속 통계 조회

* HttpRestController, ControllerTest
GET /url?key={key}
축약된 key가 저장소에 존재할 때 축약 URL에 관한 조회수 제공)

* ShortenedUrlResponse, application.yml
nullable한 필드는 json의 필드로 매핑하지 않도록 설정 변경

epicblues committed:
feat : URL 메타데이터 조회 페이지

GET /meta
메타 데이터 검색 웹 페이지(meta.html)

epicblues commented on src/main/java/com/prgrms/shortener/domain/ShortenedUrlService.java:
```diff
@@ -0,0 +1,55 @@
+    return createdUrl.getShortenedKey();
+  }
+
+  @Transactional(isolation = Isolation.READ_UNCOMMITTED)
```
갱신 대상 URL의 조회수를 최신 상태로 유지하기 위해 다른 스레드가 같은 URL을 수정한 내용을 반영하는 dirty read를 사용해 봤는데, 적합한 해결책인지 알고 싶습니다!

black9p reviewed (commented):
민성님, 과제 하시느라 고생하셨습니다.
문의내용 답변만 달았습니다.

1. 작성 안하셔도 됩니다.
2. 패키지 구성 하실때, Hexagonal architecture를 생각하신것이라면 infrastructure 패키지를 써도 됐을거 같아요.
3. 저도 주로 구현 계층이 필요 할 때, 하위패키지를 두어 사용합니다.
4. 네이밍 자체에 이상이 있어보이진 않습니다. ControllerAdvice 는 엄청 큰 규모의 프로젝트가 아니라면 하나로만 만들어 사용해도 충분합니다. 컨트롤러 별로 나눠 쓸거라면 별도로 뺄 필요없이 각 컨트롤러에서 @ExceptionHandler 만으로도 충분합니다.
5. 동시에 들어오면 다른 트랜잭션을 볼 수 있어도 같은 값을 가지고 있을 수 있기 때문에 동시성 이슈를 해결할 수 없습니다. 이 부분은 지난번 미팅때 민재님과 현정님이 조사한 방법으로 해결할 수 있을 것 같아요.
6. 어떤 이유로 이런 생각이 들었는지 싶어요. 특별히 TC를 중첩으로 구현해 사용한 기억은 없습니다.
